src/sentry/ratelimits/cardinality.py

```diff
@@ -132,3 +157,266 @@ def use_quotas(
+    sliding_windows rate limiter. You will find the next couple of paragraphs
+    in its documentation as well.
+
+    Why is checking quotas and using quotas two separate implementations?
+    Isn't that a time-of-check-time-of-use bug, and allows me to over-spend
+    quota when requests are happening concurrently?
+
+    1) It's desirable to first check quotas, then do a potentially fallible
+       operation, then consume quotas. This rate limiter is primarily going to
+       be used inside of the metrics string indexer to rate-limit database
+       writes. What we want to do there is: read DB, check rate limits,
+       write to DB, use rate limits.
+
+       If we did it any other way (the obvious one being to read DB,
+       check-and-use rate limits, write DB), crashes while writing to the
+       database can over-consume quotas. This is not a big problem if those
+       crashes are flukes, and especially not a problem if the crashes are
+       a result of an overloaded DB.
+
+       It is however a problem in case the consumer is crash-looping, or
+       crashing (quickly) for 100% of requests (e.g. due to schema
+       mismatches between code and DB that somehow don't surface during the
+       DB read). In that case the quotas would be consumed immediately and
+       incident recovery would require us to reset all quotas manually (or
+       disable rate limiting via some killswitch)
+
+    3) The redis backend (really the only backend we care about) already
+       has some consistency problems.
+
+       a) Redis only provides strong consistency and ability to
+          check-and-increment counters when all involved keys hit the same
+          Redis node. That means that a quota with prefix="org_id:123" can
+          only run on a single redis node. It also means that a global
+          quota (`prefix="global"`) would have to run on a single
+          (unchangeable) redis node to be strongly consistent. That's
+          however a problem for scalability.
+
+          There's no obvious way to make global quotas consistent with
+          per-org quotas this way, so right now it means that requests
+          containing multiple quotas with different `prefixes` cannot be
+          checked-and-incremented atomically even if we were to change the
+          rate-limiter's interface.
+
+       b) This is easily fixable, but because of the above, we
+          currently don't control Redis sharding at all, meaning that even
+          keys within a single quota's window will hit different Redis
+          node. This also helps further distribute the load internally.
+
+          Since we have given up on atomic check-and-increments in general
+          anyway, there's no reason to explicitly control sharding.
+    """
+
+    def __init__(
+        self,
+        cluster: str = "default",
+        cluster_num_shards: int = 3,
+        cluster_num_physical_shards: int = 3,
```

Comment:
please do not prefix these two argument with `cluster`.
You used `cluster` to identify the redis cluster name. `cluster_num_shards` intuitively would mean the number of shards the redis cluster has, which is not what it means.

Reply:
Also I think we can settle on having one physical shard only.  Cannot think why we would ever want to store more than one.

Reply:
I renamed the configuration parameters, you are right.

I made the number of physical shards configurable because I think we should be able to configure memory usage/sample rate separately from horizontally scaling redis, or tweaking load balance of redis nodes.

If we commit to only ever using one physical shard, and are at 10% sampling, and suddenly we observe traffic imbalance across redis nodes that forces us to increase the set sharding and/or horizontally scale, the sample rate would automatically decrease. Which may or may not be fine, but i am not sure we want to make that call if we don't have to.